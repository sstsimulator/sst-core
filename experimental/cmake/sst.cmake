# Set variables for sst_config.h attempt to just go in order
include(CheckSymbolExists)
include(CheckIncludeFile)
include(CheckLibraryExists)

check_symbol_exists(argz_add "argz.h" HAVE_ARGZ_ADD)
check_symbol_exists(argz_append "argz.h" HAVE_ARGZ_APPEND)
check_symbol_exists(argz_count "argz.h" HAVE_ARGZ_COUNT)
check_symbol_exists(argz_create_sep "argz.h" HAVE_ARGZ_CREATE_SEP)
check_include_file(argz.h HAVE_ARGZ_H)
check_symbol_exists(argz_insert "argz.h" HAVE_ARGZ_INSERT)
check_symbol_exists(argz_next "argz.h" HAVE_ARGZ_NEXT)
check_symbol_exists(argz_stringify "argz.h" HAVE_ARGZ_STRINGIFY)
check_symbol_exists(closedir "sys/types.h" HAVE_CLOSEDIR)
check_include_file(c_asm.h HAVE_C_ASM_H)
check_symbol_exists(cygwin_conv_path "sys/cygwin.h" HAVE_DECL_CYGWIN_CONV_PATH)
check_include_file(dirent.h HAVE_DIRENT_H)
check_library_exists(dld dld_init "" HAVE_DLD)
check_include_file(dld.h HAVE_DLD_H)
check_symbol_exists(dlerror "dlfcn.h" HAVE_DLERROR)
check_include_file(dlfcn.h HAVE_DLFCN_H)
check_include_file(dl.h HAVE_DL_H)
check_symbol_exists(_dyld_func_lookup "dyld.h" HAVE_DYLD)
check_symbol_exists(error_t "errno.h" HAVE_ERROR_T)

if(HDF5_FOUND)
  set(HAVE_HDF5 ON)
endif(HDF5_FOUND)

check_include_file(intrinsics.h HAVE_INTRINSICS_H)
check_include_file(inttypes.h HAVE_INTTYPES_H)
if(HAVE_INTTYPES_H)
  # TODO do we need to check that symbols like PRIu64 exist first?
  set(__STDC_FORMAT_MACROS ON)
endif(HAVE_INTTYPES_H)

if(CMAKE_DL_LIBS)
  set(HAVE_LIBDL ON)
endif()

if(ZLIB_FOUND)
  set(HAVE_LIBZ ON)
endif(ZLIB_FOUND)

if(CURSES_FOUND)
  set(HAVE_CURSES ON)
endif(CURSES_FOUND)

check_include_file(execinfo.h HAVE_EXECINFO_H)
check_symbol_exists(backtrace "execinfo.h" HAVE_BACKTRACE)
check_include_file(mach/mach_time.h HAVE_MACH_MACH_TIME_H)
check_include_file(mach-o/dyld.h HAVE_MACH_O_DYLD_H)
check_symbol_exists(opendir "dirent.h" HAVE_OPENDIR)
check_symbol_exists(readdir "dirent.h" HAVE_READDIR)

# TODO we really should fix how we do c++ std flags
if(CMAKE_CXX_STANDARD GREATER_EQUAL 14)
  set(HAVE_STDCXX_1Y ON)
endif()

check_include_file(stdint.h HAVE_STDINT_H)
check_include_file(stdio.h HAVE_STDIO_H)
check_include_file(stdlib.h HAVE_STDLIB_H)
check_include_file(strings.h HAVE_STRINGS_H)
check_include_file(string.h HAVE_STRING_H)
check_symbol_exists(strlcat "string.h" HAVE_STRLCAT)
check_symbol_exists(strlcpy "string.h" HAVE_STRLCPY)
check_include_file(sys/dl.h HAVE_SYS_DL_H)
check_include_file(sys/stat.h HAVE_SYS_STAT_H)
check_include_file(sys/stat.h HAVE_SYS_STAT_H)
check_include_file(sys/time.h HAVE_SYS_TIME_H)
check_include_file(sys/types.h HAVE_SYS_TYPES_H)
check_include_file(unistd.h HAVE_UNISTD_H)

if(HAVE_ARGZ_H)
  set(HAVE_WORKING_ARGZ ON)
endif()

set(PACKAGE_BUGREPORT "https://github.com/sstsimulator/sst-core/issues")
set(PACKAGE_NAME "SSTCore")
set(PACKAGE_NAME "https://github.com/sstsimulator/sst-core")

set(SST_CC ${CMAKE_C_COMPILER})
set(SST_CFLAGS ${CMAKE_C_FLAGS})

if(APPLE)
  set(SST_COMPILE_MACOSX 1)
endif()

if(MPI_FOUND)
  set(SST_CONFIG_HAVE_MPI ON)
  set(SST_MPICC ${MPI_C_COMPILER})
  set(SST_MPICXX ${MPI_CXX_COMPILER})
endif()

if(Python_FOUND)
  set(HAVE_PYTHON_H 1)
  set(SST_CONFIG_HAVE_PYTHON ON)
  # CMAKE build can just require PYTHON3
  set(SST_CONFIG_HAVE_PYTHON3 ON)
  set(SST_PTYHON_LDFLAGS "${Python_LIBRARIES}")
  set(SST_PYTHON_CPPFLAGS "-I${Python_INCLUDE_DIRS}")

  # NOTE I think we should try to avoid using python-config if possible
  # find_program(PYCONFIG NAMES python-config "${Python_EXECUTABLE}-config")
  # if(PYCONFIG) message(STATUS "Found PYCONFIG ${PYCONFIG}.")
  #
  # #execute_process( #  COMMAND ${PYCONFIG} --ldflags #  RESULT_VARIABLE
  # PYCONFIG_RESULT #  OUTPUT_VARIABLE SST_PYTHON_LDFLAGS #
  # OUTPUT_STRIP_TRAILING_WHITESPACE)
  #
  # #if(PYCONFIG_RESULT AND NOT PYCONFIG_RESULT EQUAL 0) #  message( #
  # FATAL_ERROR #      "python-config (${PYCONFIG}) --ldflags failed with
  # ${PYCONFIG_RESULT}") #endif(PYCONFIG_RESULT AND NOT PYCONFIG_RESULT EQUAL 0)
  # set(SST_PTYHON_LDFLAGS ${Python_LIBRARIES})
  #
  # execute_process( COMMAND ${PYCONFIG} --cppflags RESULT_VARIABLE
  # PYCONFIG_RESULT OUTPUT_VARIABLE SST_PYTHON_CPPFLAGS
  # OUTPUT_STRIP_TRAILING_WHITESPACE) else(PYCONFIG) message(FATAL_ERROR "Failed
  # to find python-config") endif(PYCONFIG)
endif()

set(SST_CPP ${CMAKE_CXX_COMPILER})
set(SST_CPPFLAGS ${CMAKE_CXX_FLAGS})
set(SST_CXXCPP ${CMAKE_CXX_COMPILER})
set(SST_CXX ${CMAKE_CXX_COMPILER})
set(SST_CXXFLAGS ${CMAKE_CXX_FLAGS})
set(SST_INSTALL_PREFIX ${CMAKE_INSTALL_PREFIX})
set(SST_LD ${CMAKE_LINKER})
set(SST_LDFLAGS "${CMAKE_CXX_LINK_FLAGS} ${LINK_FLAGS}")

# This is a workaround for sstsimulator.conf LD, the autotools expects LIBS to
# be -l lib but if CMAKE_DL_LIBS is empty you just get a -l whic breaks linking
if(CMAKE_DL_LIBS)
  set(SST_DL_LIBS "-l${CMAKE_DL_LIBS}")
endif()

if(NOT SST_DISABLE_MEM_POOLS)
  set(USE_MEMPOOL ON)
endif()

set(SST_BUILD_WITH_CMAKE ON)
set(PACKAGE_VERSION ${CMAKE_PROJECT_VERSION})

if(Python_VERSION AND Python_VERSION_MAJOR GREATER_EQUAL 3)
  set(SST_CONFIG_HAVE_PYTHON3 ON)
  set(HAVE_PYTHON_H ON)
endif()

find_package(Git REQUIRED)
execute_process(
  COMMAND ${GIT_EXECUTABLE} --git-dir=${SST_TOP_SRC_DIR}/.git rev-parse HEAD
  RESULT_VARIABLE HASH_RESULT
  OUTPUT_VARIABLE SSTCORE_GIT_HEADSHA
  OUTPUT_STRIP_TRAILING_WHITESPACE)

execute_process(
  COMMAND ${GIT_EXECUTABLE} --git-dir=${SST_TOP_SRC_DIR}/.git rev-parse
          --abbrev-ref HEAD
  RESULT_VARIABLE BRANCH_RESULT
  OUTPUT_VARIABLE SSTCORE_GIT_BRANCH
  OUTPUT_STRIP_TRAILING_WHITESPACE)

check_include_file(sys/stat.h HAVE_SYS_STAT_H)
check_include_file(dlfcn.h HAVE_DLFCN_H)
check_include_file(sys/types HAVE_SYS_TYPES_H)
check_include_file(unistd.h HAVE_UNISTD_H)

check_library_exists(m sin "" HAVE_LIBM)
if(NOT HAVE_LIBM)
  message(FATAL_ERROR "Failed to detect libm")
endif()

check_library_exists(rt aio_cancel "" HAVE_LIBRT)
