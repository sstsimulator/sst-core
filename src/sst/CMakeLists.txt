# ~~~
# SST-CORE src/sst/ CMake
#
# Copyright 2009-2023 NTESS. Under the terms
# of Contract DE-NA0003525 with NTESS, the U.S.
# Government retains certain rights in this software.
#
# Copyright (c) 2009-2023, NTESS
# All rights reserved.
#
# This file is part of the SST software package. For license
# information, see the LICENSE file in the top level directory of the
# distribution.
# ~~~
#

add_subdirectory(core)
