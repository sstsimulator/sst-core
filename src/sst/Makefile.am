# Copyright 2009-2024 NTESS. Under the terms
# of Contract DE-NA0003525 with NTESS, the U.S.
# Government retains certain rights in this software.
#
# Copyright (c) 2009-2024, NTESS
# All rights reserved.
#
# This file is part of the SST software package. For license
# information, see the LICENSE file in the top level directory of the
# distribution.

SUBDIRS = \
	core

pkgconfigdir = $(libdir)/pkgconfig
pkgconfig_DATA = SST-${PACKAGE_VERSION}.pc
