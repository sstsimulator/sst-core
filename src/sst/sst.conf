# ===================================================================
# SST Configuration Database
#
# This file is automatically generated during installation. Any
# changes made manually may be overridden. Please see the SST
# documentation for details on how the configurations should be
# edited to prevent errors during execution
#
# ===================================================================
[SSTCore]
MPICXX=@MPICXX@
MPICC=@MPICC@
MPI_CPPFLAGS=@MPI_CPPFLAGS@
CXX=@CXX@
CXXFLAGS=@SST_EXPORT_CXXFLAGS@
CC=@CC@
CFLAGS=@SST_EXPORT_CFLAGS@
LD=@LD@
LDFLAGS=@LDFLAGS@
LIBS=@LIBS@
CPP=@CPP@
CXXCPP=@CXXCPP@
AR=@AR@
NM=@NM@
PYTHON=@PYTHON_EXE@
PYTHON_CONFIG=@PYTHON_CONFIG_EXE@
PYTHON_CPPFLAGS=@PYTHON_CPPFLAGS@
PYTHON_LDFLAGS=@PYTHON_LDFLAGS@
PYTHON_VERSION3=@PYTHON_VERSION3@
ZOLTAN_CPPFLAGS=@ZOLTAN_CPPFLAGS@
ZOLTAN_LDFLAGS=@ZOLTAN_LDFLAGS@
LIBZ_CPPFLAGS=@LIBZ_CPPFLAGS@
LIBZ_LDFLAGS=@LIBZ_LDFLAGS@
LIBZ_LIBS=@LIBZ_LIBS@
ELEMENT_CXXFLAGS=@SST_EXPORT_CXXFLAGS@ @SST_ELEMENT_FPIC_FLAGS@ -DHAVE_CONFIG_H -I@prefix@/include
ELEMENT_LDFLAGS=-shared -fno-common -Wl,-undefined -Wl,dynamic_lookup
pkgconfig=@prefix@/lib/pkgconfig/SST-@PACKAGE_VERSION@.pc
sourcedir=@abs_srcdir@/core
testsdir=@abs_srcdir@/../../tests
prefix=@prefix@
exec_prefix=@prefix@
bindir=@prefix@/bin
sbindir=@prefix@/sbin
libexecdir=@prefix@/libexec
includedir=@prefix@/include
libdir=@prefix@/lib
CORE_TEST_ELEMENT_LIBDIR=@prefix@/lib/sstcore
PREVIEW_BUILD=@SST_ENABLE_PREVIEW_BUILD_STR@
