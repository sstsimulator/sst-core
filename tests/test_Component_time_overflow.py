import sst

# Define SST core options
#sst.setProgramOption("stop-at", "25us")

# Define the simulation components
comp_c0_0 = sst.Component("c0_0", "coreTestElement.coreTestComponent")
comp_c0_0.addParams({
    "workPerCycle" : "100",
    "commSize" : "100",
    "commFreq" : "1000",
    "clockFrequency" : "60s"
})

comp_c0_1 = sst.Component("c0_1", "coreTestElement.coreTestComponent")
comp_c0_1.addParams({
    "workPerCycle" : "100",
    "commSize" : "100",
    "commFreq" : "1000",
    "clockFrequency" : "60s"
})

comp_c1_0 = sst.Component("c1_0", "coreTestElement.coreTestComponent")
comp_c1_0.addParams({
    "workPerCycle" : "100",
    "commSize" : "100",
    "commFreq" : "1000",
    "clockFrequency" : "60s"
})

comp_c1_1 = sst.Component("c1_1", "coreTestElement.coreTestComponent")
comp_c1_1.addParams({
    "workPerCycle" : "100",
    "commSize" : "100",
    "commFreq" : "1000",
    "clockFrequency" : "60s"
})

# Define the simulation links

# North/South links
link_ns_0_01 = sst.Link("link_ns_0_01")
link_ns_0_01.connect( (comp_c0_0, "Nlink", "10000ps"), (comp_c0_1, "Slink", "10000ps") )

link_ns_0_10 = sst.Link("link_ns_0_10")
link_ns_0_10.connect( (comp_c0_0, "Slink", "10000ps"), (comp_c0_1, "Nlink", "10000ps") )

link_ns_1_01 = sst.Link("link_ns_1_01")
link_ns_1_01.connect( (comp_c1_0, "Nlink", "10000ps"), (comp_c1_1, "Slink", "10000ps") )

link_ns_0_10 = sst.Link("link_ns_1_10")
link_ns_0_10.connect( (comp_c1_0, "Slink", "10000ps"), (comp_c1_1, "Nlink", "10000ps") )

# East/West links
link_ew_0_01 = sst.Link("link_ew_0_01")
link_ew_0_01.connect( (comp_c0_0, "Elink", "10000ps"), (comp_c1_0, "Wlink", "10000ps") )

link_ew_0_10 = sst.Link("link_ew_0_10")
link_ew_0_10.connect( (comp_c0_0, "Wlink", "10000ps"), (comp_c1_0, "Elink", "10000ps") )

link_ew_1_01 = sst.Link("link_ew_1_01")
link_ew_1_01.connect( (comp_c0_1, "Elink", "10000ps"), (comp_c1_1, "Wlink", "10000ps") )

link_ew_1_10 = sst.Link("link_ew_1_10")
link_ew_1_10.connect( (comp_c0_1, "Wlink", "10000ps"), (comp_c1_1, "Elink", "10000ps") )
