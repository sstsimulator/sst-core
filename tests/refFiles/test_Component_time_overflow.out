FATAL:  SST Core: ERROR: SST Core detected a time fault (an event had an earlier time than the previous event). The most likely cause of this is that the 64-bit core time had an overflow condition.  This is typically caused by having low frequency events with too low of a timebase.  See the extended help for --timebase option (sst --help timebase)
SST Fatal Backtrace Information:
    0 : 0   sstsim.x                            0x000000010024ea3f _ZNK3SST6Output5fatalEjPKcS2_iS2_z + 879
    1 : 1   sstsim.x                            0x0000000100269b82 _ZN3SST15Simulation_impl3runEv + 866
    2 : 2   sstsim.x                            0x00000001001f1b20 _ZL16start_simulationjR15SimThreadInfo_tRN3SST4Core10ThreadSafe7BarrierE + 3200
    3 : 3   sstsim.x                            0x00000001001ed044 main + 7492
    4 : 4   dyld                                0x000000010916852e start + 462
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 5.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
